articles/141_static_remapping.md

```diff
@@ -0,0 +1,341 @@
+Client libraries also have APIs in code to pass remapping rules when the node is initialized.
+A remap rule consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+Both sides of a rule are [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b).
+Before a name is remapped it is also [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f).
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+
+## Remapping rule syntax
+This is a proposal for the ROS 2 remapping rule syntax.
+It attempts to be the same as ROS 1 syntax when possible, and different where it supports new features.
+
+Requirements supported by this syntax
+- Remap One Node in a Process
+- Change a Namespace
+- Change a Basename
+- Exact FQN Replacement
+- Exact Relative Name Replacement
+- Remap via Command Line
+- Change the Default Namespace
+
+Requirements not supported
+- Change a Token
+- Pre-FQN Remapping
+
+Special Operators
+- `*` wild card that matches a single token
+- `**` wild card that matches at least one slash and zero or more tokens
+- `:=` divides the two parts of a remapping rule: matching and replacement
+- `~` expands to `/namespace/nodename` and may only be at the very beginning of a match or replacement
+- `__ns` matches the default namespace if used by itself on the match side of a rule
+- `nodename:` prefixed to a rule makes it apply only to a node with that name
+- `\1` - `\9` are replaced with the matched content of a wild card
+
+The operators `*` and `**` are chosen to match globbing behavior in bash.
+`**` behaves similarly as it does in bash>=4.0 with the globstar option set.
+The syntax for `\[1-9]` was taken from the syntax for backreferences in POSIX BRE.
+The operators `~`, `:=`, and `__ns` have the same meaning in ROS 1.
+
+### Applications of the syntax
+The following sections explain how the syntax enables the use cases above.
+
+#### Supporting: Remap One Node in a Process
+Remapping a node in a process requires a way to uniquely identify a node.
+Assuming the node's name is unique in a process, a rule can be prefix with the name and a `:`.
+If the node name is not prefixed, the rule should be applied to all nodes in the process.
+
+*Example*
+
+- Multiple nodes in a process use the name `scan`
+- A rule is given to one node `node1:scan:=scan_filtered`
+- Only node1 uses the rule
+
+
+#### Supporting: Change a Namespace
+There are two cases: changing part of a namespace, and changing the entire namespace.
+The first case requires a wildcard to match the rest of a namespace.
+The second requires a wildcard to match the basename at the end.
+
+*Example: Partial namespace replacement*
+
+- Node uses names `/foo`, `/foo/bar/baz`
+- Node given rule `/foo**:=/fizz\1`
```

Comment:
I agree that `/foo/**` looks better. I think that if you wanted to make the existing `/foo**` make sense, you'd have to have an input topic like `/foo_bar/baz` which would become `/fizz_bar/baz`. However, I think to do what @dirk-thomas suggested, i.e. `/foo/**`, and have it match `/foo/bar` (for example) then the above rule about `**` would need to be modified to so that it could match zero to many slashes rather than at least one.

At that point, however, it's hard to make a rule to match like this:

- `/foo`: does not match
- `/foo/bar`: matches, `/1` is `/bar` for `/foo**` or `bar` for `/foo/**`
- `/foo/bar/baz`: matches, `/1` is `/bar/baz` for `/foo**` or `bar/baz` for `/foo/**`

The current setup `/foo**` would work in most of these cases (except `/1` is `/bar` for `/foo**`), but if you changed it so `**` doesn't have to match at least one `/`, then it would also match `/foo`. I'm not sure it matters which, but you'll have to choose between a more flexible `**` and a stricter `**` which might give you more control.